### PR TITLE
Sparse

### PR DESCRIPTION
**Pull request by kedz**
I added a SparseSklearnIntentClassifier model and a simple test fixture/test config. The model code largely comes from the SklearnIntentClassifier from rasa OS, but swaps out the dense features for sparse features. Additionally, instead of using a support vector machine based classifier, it uses the BernoulliNB model in sklearn.

I'm not sure how extensively this needs to be tested, but I could also try to emulate what is done for the DIETClassifier (as far as I can tell, SklearnIntentClassifier in rasa OS doesn't actually have any tests?). Right now the one test just verifies that SparseSklearnIntentClassifier object produces identical predictions to those of BernoulliNB when given the same training data/hyperparameters.

### CODE REVIEW

File: rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+from rasa.shared.nlu.training_data.training_data import TrainingData
+from rasa.shared.nlu.training_data.message import Message
+
+logger = logging.getLogger(__name__)
+
+if typing.TYPE_CHECKING:
+    import sklearn
+
+
+class SparseSklearnIntentClassifier(IntentClassifier):
```
Comment (koaning) [inline rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py:29]: Technically we may be implementing more sparse classifiers from sklearn. Maybe rename to `SparseNaiveBayesClassifier`?

File: tests/configs/sparse-sklearn-intent-classifier-config.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+pipeline:
+- name: WhitespaceTokenizer
+- name: CountVectorsFeaturizer
+- name: CountVectorsFeaturizer
+  analyzer: char_wb
+  min_ngram: 1
+  max_ngram: 4
+- name: rasa_nlu_examples.classifiers.SparseSklearnIntentClassifier
+  alpha: [0.1]
```
Comment (koaning) [inline tests/configs/sparse-sklearn-intent-classifier-config.yml:11]: Shouldn't this be a float instead of a list of floats?

File: tests/test_classifiers/test_sparse_sklearn_intent_classifier.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def test_predict():
+    NLU_DATA_PATH = "tests/data/nlu/en/nlu.md"
+    model_path = train_nlu(
+        nlu_data=NLU_DATA_PATH,
+        config="tests/configs/sparse-sklearn-intent-classifier-config.yml",
+        output="models",
+    )
+
```
Comment (koaning) [inline tests/test_classifiers/test_sparse_sklearn_intent_classifier.py:25]: Could you also add a smoketest? 

https://github.com/RasaHQ/rasa-nlu-examples/blob/master/tests/test_smoke/test_english.py#L15

**Review by koaning — COMMENTED** [review]
Some minor comments for friday evening. Will re-review Monday! On the whole: it's looks good 👍

**Comment by sara-tagger** [comment]
Thanks for submitting a pull request 🚀 @dakshvar22 will take a look at it as soon as possible ✨

File: tests/test_classifiers/test_sparse_sklearn_intent_classifier.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import pathlib
+
```
Comment (koaning) [inline tests/test_classifiers/test_sparse_sklearn_intent_classifier.py:2]: What happens if we add dense data to our pipeline too? We should raise a proper error message if sklearn breaks.

File: rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                intent_ranking = []
+
+        message.set("intent", intent, add_to_output=True)
+        message.set("intent_ranking", intent_ranking, add_to_output=True)
+
+    def predict_prob(self, X: scipy.sparse.spmatrix) -> np.ndarray:
+        """Given a bow vector of an input text, predict the intent label.
+        Return probabilities for all labels.
+        :param X: bow of input text
+        :return: vector of probabilities containing one entry for each label"""
```
Comment (koaning) [inline rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py:215]: Nitpick. But doens't the styleguide prefer

```
"""
Given a bow vector of an input text, predict the intent label.
        Return probabilities for all labels.
        :param X: bow of input text
        :return: vector of probabilities containing one entry for each label
"""
```

instead of 

```
"""Given a bow vector of an input text, predict the intent label.
        Return probabilities for all labels.
        :param X: bow of input text
        :return: vector of probabilities containing one entry for each label"""
```

File: rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        # str not instance of basestr...
+        tuned_parameters = [
+            {"alpha": alpha, "binarize": binarize, "fit_prior": fit_prior}
+        ]
+
+        # aim for 5 examples in each fold
+
+        cv_splits = self._num_cv_splits(y)
+
+        return GridSearchCV(
```
Comment (koaning) [inline rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py:168]: Ah. Now I understand the list in the `config.yml` file. Do we really want to use a GridSearch here? Rasa has GridSearch functionality via the command line and we also have a [hyperopt](https://github.com/RasaHQ/nlu-hyperopt) library. 
With that in mind it feels like overkill to *also* do GridSearch inside of a model.

File: rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+                self._get_sentence_features(example)
+                for example in training_data.intent_examples
+            ]
+        )
+
+        return X, y
+
+    @staticmethod
+    def _get_sentence_features(message: Message) -> scipy.sparse.spmatrix:
+        _, sentence_features = message.get_sparse_features(TEXT)
```
Comment (koaning) [inline rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py:137]: What happens if we add dense data to our pipeline too? It might be good to raise a warning.

File: rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+        if sentence_features is not None:
+            return sentence_features.features
+
+        raise ValueError(
+            "No sentence features present. Not able to train sklearn policy."
```
Comment (koaning) [inline rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py:142]: Maybe better:

```
"No sparse sentence features present. Not able to train sklearn policy."
```

**Comment by koaning** [comment]
I suppose a final thing that might be interesting to do (totally not required), have you tried running this model against a meaningful bot? It'd be a "nice-to-have" benchmark.

File: rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py (hunk 1 above)
Comment (kedz) [reply rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py:29]: Yeah I can change the name. One thing I'm hazy on though, I think most scikit learn models can handle sparse data or dense data. Why not make a NaiveBayesClassifier, and give it an option flag to select whether to use sparse or dense features from the rasa feature extraction pipeline?

File: rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py (hunk 6 above)
Comment (kedz) [reply rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py:168]: Yeah that makes sense to me to remove grid search. The grid search was what was done in the SklearnIntentClassifier from the the rasa OS code base and I was just trying to make minimal changes to that. Will remove it.

**Comment by kedz** [comment]
> I suppose a final thing that might be interesting to do (totally not required), have you tried running this model against a meaningful bot? It'd be a "nice-to-have" benchmark.

Yeah I would be interested in doing this. Do you have any pointers to where/how previous rasa nlu components have been benchmarked?

File: rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py (hunk 7 above)
Comment (kedz) [reply rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py:137]: How does this warning message strike you, "Dense features are being computed but not used in the SparseNaiveBayesIntentClassifier." ?

**Comment by koaning** [comment]
I've got a benchmarking guide [here](https://rasahq.github.io/rasa-nlu-examples/benchmarking/) but it is fairly general. I'm not aware of a central place where we store all of our benchmarks though. It is however on my road-map to "do something with". We might include a `benchmarks.md` file to the central Rasa repo with some stats to also help others compare. 

It's a open problem though, but feel free to ping the slack channel if you've got ideas on that.

File: rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py (hunk 7 above)
Comment (koaning) [reply rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py]: That sounds perfect :)

File: rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py (hunk 1 above)
Comment (koaning) [reply rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py]: Scikit-learn offers [many different kinds of NaiveBayes models](https://scikit-learn.org/stable/modules/naive_bayes.html) actually. I even maintain an open-source project that adds [yet another kind](https://scikit-lego.readthedocs.io/en/latest/naive-bayes.html) too. Scikit-learn implicitly translates sparse data whenever it is combined with dense data into dense data. It's a fair way of dealing with it, but it's not very performant. 

As far as Rasa users are concerned though, I think it's fair to assume that DIET will always be more performant in terms of accuracy. The main use-case for a sparse model here is to have a simple benchmark that is *much* faster to train. I'm currently trying to implement a form and I need to wait for 10-20s every time I iterate because DIET needs to train. It'd be nice if this was reduced.

**Comment by kedz** [comment]
Ok aside from the benchmarking which is on my todos, I think I have addressed your comments, but lemme know if there's anything else that needs fixing.

**Comment by koaning** [comment]
LGTM.

The only comment is that documentation is missing. If you're swamped just tell me and I'll take care of it though.

**Comment by koaning** [comment]
Ah! 

I found a bug in the CI. I recently switched from `master` to `main` and that means that currently your branch isn't being tested here. I'll make a push to master with a fix and merge here after.

**Comment by kedz** [comment]
@koaning Just added docs which you should check out. Regarding merge into master/main, I'm not sure I understand. What do you want me to do exactly?

**Comment by koaning** [comment]
If your the branch owner then usually there's a "sync master" button here.

image

That's what I was referring to. That said -> I'll merge and take it from here.

Thanks for the PR!
